src/widgets/declarative/DeclarativeWidget.ts

```diff
@@ -0,0 +1,245 @@
+   */
+  analyticsApi: AnalyticsApi;
+
+  /**
+   * Instance of {@link EmbedWidget} or {@link PopupWidget}
+   */
+  widgetInstance: EmbedWidget | PopupWidget;
+
+  /**
+   * Determines whether to render the widget as an embedding widget or popup widget.
+   */
+  type: "EMBED" | "POPUP";
+
+  /**
+   * Container element to contain the widget iframe
+   * @default this
+   */
+  container: string | HTMLElement | undefined | null;
+  element: HTMLElement | undefined;
+
+  constructor() {
+    super();
+    this.attachShadow({
+      mode: "open",
+    }).innerHTML = `<style>:host { display: contents; }</style><slot></slot>`;
+
+    this.config = getConfig();
+    this.token = getToken();
+    this.tenant = window.squatchTenant;
+    this.container = this;
+  }
+
+  private _setupApis(config?: ConfigOptions) {
+    if (!this.tenant) throw new Error("tenantAlias not provided");
+
+    this.widgetApi = new WidgetApi({
+      tenantAlias: config?.tenantAlias || this.tenant,
+      domain: config?.domain || this.config?.domain || DEFAULT_DOMAIN,
+    });
+    this.analyticsApi = new AnalyticsApi({
+      domain: config?.domain || this.config?.domain || DEFAULT_DOMAIN,
+    });
+  }
+
+  private async renderPasswordlessVariant() {
+    this._setupApis();
+
+    _log("Rendering as an Instant Access widget");
+
+    return await this.widgetApi
+      .render({
+        engagementMedium: this.type,
+        widgetType: this.widgetType,
+        locale: this.locale,
+      })
+      .then((res) => this._setWidget(res.template, { type: "passwordless" }))
+      .catch(this.setErrorWidget);
+  }
+
+  private async renderUserUpsertVariant() {
+    this._setupApis();
+
+    const userObj = decodeUserJwt(this.token!);
+    if (!userObj) {
+      return this.setErrorWidget(Error("No user object in token."));
+    }
+
+    _log("Rendering as a Verified widget");
+
+    const widgetInstance = await this.widgetApi
+      .upsertUser({
+        user: userObj,
+        locale: this.locale,
+        engagementMedium: this.type,
+        widgetType: this.widgetType,
+        jwt: this.token,
+      })
+      .then((res) =>
+        this._setWidget(res.template, { type: "upsert", user: userObj })
+      )
+      .catch(this.setErrorWidget);
+
+    return widgetInstance;
+  }
+
+  private _setWidget = (
+    template: any,
+    config: { type: "upsert" | "passwordless"; user?: User }
+  ) => {
+    const params = {
+      api: this.widgetApi,
+      content: template,
+      context: {
+        type: config.type,
+        user: config.user,
+        container: this.container || this,
+        engagementMedium: this.type,
+      },
+      type: this.widgetType!,
+      domain: this.config?.domain || DEFAULT_DOMAIN,
+      npmCdn: DEFAULT_NPM_CDN,
+    };
+    if (this.type === "EMBED") {
+      return new EmbedWidget(params, params.context.container);
+    } else {
+      return new PopupWidget(params, this.firstChild ? null : undefined);
```

Comment:
```suggestion
      return new PopupWidget(params, this.firstChild ? null : undefined);
```
this is a weird ternary, why are null and undefined treated differently?

Reply:
It's to have `PopupWidget` set a default trigger to `.squatchpop` if there is no firstChild. If there is, the trigger should be `null` since we listen to a click event from the children of the web component instead of a specific trigger.

Reply:
ah that makes sense, it might help to write this in a bit more self-documenting way then, something like 
```suggestion
        const useFirstChildTrigger = this.firstChild ? null : undefined;
        return new PopupWidget(params, useFirstChildTrigger);
```